### ISSUE TITLE
From  pandas to MolDataset with batch

### ISSUE THREAD

**Issue by thegodone** [mlx-graphs/mlx-graphs#171]
**Is your feature request related to a problem? Please describe.**

be able to do something like this: 

````
    df = pd.DataFrame({
        'smiles': ['CCO', 'CCN', 'CCC', 'C1=CC=CC=C1'],
        'target': [0, 1, 0, 1]
    })

    dataset = SMILESDataset(df, use_edges=True, use_partial_charge=True, cache_dir="graph_cache")

    dataset.process()

    dataset.set_for_dataloader(True)

    train_loader = Dataloader(dataset, batch_size=2, shuffle=False)

    for batch in train_loader:
        print(batch.node_features)
        print(batch.edge_index)
        print(batch.edge_features)
````

**Describe the solution you'd like**
````
import numpy as np
from rdkit.Chem.rdPartialCharges import ComputeGasteigerCharges
from rdkit import Chem
from mlx_graphs.data.data import GraphData  # Keep using GraphData from mlx_graphs
from typing import List, Tuple, Union, Optional


# One-hot encoding utility
def one_hot_encode(
    val: Union[int, str], 
    allowable_set: Union[List[str], List[int]], 
    include_unknown_set: bool = False
) -> List[float]:
    """One hot encoder for elements of a provided set."""
    one_hot = [0.0] * (len(allowable_set) + int(include_unknown_set))
    if val in allowable_set:
        one_hot[allowable_set.index(val)] = 1.0
    elif include_unknown_set:
        one_hot[-1] = 1.0
    return one_hot


# Get atom partial charge
def get_atom_partial_charge(atom: Chem.Atom) -> List[float]:
    """Get the partial charge of an atom from RDKit Mol object."""
    gasteiger_charge = atom.GetProp("_GasteigerCharge")
    if gasteiger_charge in ["-nan", "nan", "-inf", "inf"]:
        gasteiger_charge = 0.0
    return [float(gasteiger_charge)]


# Construct atom feature
def _construct_atom_feature(
    atom: Chem.Atom,
    use_partial_charge: bool,
    allowable_set: List[str],
) -> np.ndarray:
    """Construct a feature vector for a single atom."""
    atom_type = one_hot_encode(atom.GetSymbol(), allowable_set, True)
    formal_charge = [float(atom.GetFormalCharge())]
    hybridization = one_hot_encode(str(atom.GetHybridization()), ["SP", "SP2", "SP3"], False)
    aromatic = [float(atom.GetIsAromatic())]
    degree = one_hot_encode(atom.GetTotalDegree(), [0, 1, 2, 3, 4, 5], True)
    
    atom_feat = np.concatenate([atom_type, formal_charge, hybridization, aromatic, degree])

    if use_partial_charge:
        partial_charge = get_atom_partial_charge(atom)
        atom_feat = np.concatenate([atom_feat, np.array(partial_charge)])

    return atom_feat


# Construct bond feature
def _construct_bond_feature(bond: Chem.Bond) -> np.ndarray:
    """Construct a feature vector for a single bond."""
    bond_type = one_hot_encode(str(bond.GetBondType()), ["SINGLE", "DOUBLE", "TRIPLE", "AROMATIC"], False)
    same_ring = [int(bond.IsInRing())]
    conjugated = [int(bond.GetIsConjugated())]
    stereo = one_hot_encode(str(bond.GetStereo()), ["STEREONONE", "STEREOANY", "STEREOZ", "STEREOE"], True)
    return np.concatenate([bond_type, same_ring, conjugated, stereo])


# Molecular Graph Featurizer
```
class MolGraphConvFeaturizer:
    def __init__(self, use_edges: bool = False, use_partial_charge: bool = False):
        self.use_edges = use_edges
        self.use_partial_charge = use_partial_charge

    def featurize(self, datapoint: Chem.Mol, allowable_set: List[str]) -> GraphData:
        """
        Featurize a molecule represented by an RDKit Mol object into a graph format (node/edge features).

        Parameters
        ----------
        datapoint: RDKit Mol object.
        allowable_set: List of atoms you want a feature for in the atom feature vector.

        Returns
        -------
        GraphData: A graph data structure containing node features, edge index, and optionally edge features.
        """
        # Ensure more than one atom in molecule
        assert datapoint.GetNumAtoms() > 1, "Molecule should have more than one atom."

        # Compute partial charges if needed
        if self.use_partial_charge:
            try:
                datapoint.GetAtomWithIdx(0).GetProp("_GasteigerCharge")
            except KeyError:
                ComputeGasteigerCharges(datapoint)

        # Construct node features (atoms)
        atom_features = np.asarray(
            [
                _construct_atom_feature(atom, self.use_partial_charge, allowable_set)
                for atom in datapoint.GetAtoms()
            ], dtype=float
        )

        # Construct edge index (bonds)
        src, dest = [], []
        for bond in datapoint.GetBonds():
            start, end = bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()
            src += [start, end]
            dest += [end, start]

        # Construct edge features (bonds)
        bond_features = None
        if self.use_edges:
            bond_features = np.asarray(
                [
                    _construct_bond_feature(bond)
                    for bond in datapoint.GetBonds()
                ] * 2, dtype=float  # Double for both directions
            )

        # Return the molecular graph data using the existing mlx.data.GraphData
        return GraphData(
            edge_index=np.asarray([src, dest], dtype=int),
            node_features=atom_features,
            edge_features=bond_features if self.use_edges else None,
        )

if __name__ == "__main__":

    # Example SMILES string
    smiles = "C1CCCCC1"  # Cyclohexane
    mol = Chem.MolFromSmiles(smiles)

    # Initialize the featurizer with edge features
    featurizer = MolGraphConvFeaturizer(use_edges=True)

    # Featurize the molecule
    graph_data = featurizer.featurize(mol,  allowable_set=["C", "O", "N", "H"])

    # Output the graph data
    print(graph_data)

```
and a MolDataset similar to this:
```

import os
import pickle
import pandas as pd
import numpy as np
from rdkit import Chem
import mlx.core as mx
from mlx_graphs.data.data import GraphData
from mlx_graphs.datasets import Dataset
from featurizer import MolGraphConvFeaturizer
from typing import Union, Sequence, Optional
from mlx_graphs.loaders import Dataloader


def save_graph_data(graph_data: GraphData, file_path: str):
    with open(file_path, 'wb') as f:
        pickle.dump(graph_data, f)


def load_graph_data(file_path: str) -> GraphData:
    with open(file_path, 'rb') as f:
        return pickle.load(f)


def smiles_to_graph(smiles: str, featurizer: MolGraphConvFeaturizer, allowable_set: list) -> GraphData:
    mol = Chem.MolFromSmiles(smiles)
    if mol is None:
        raise ValueError(f"Could not parse SMILES string: {smiles}")

    graph_data = featurizer.featurize(mol, allowable_set=allowable_set)

    assert isinstance(graph_data.node_features, np.ndarray), "Node features should be numpy arrays"
    assert graph_data.edge_features is None or isinstance(graph_data.edge_features, np.ndarray), "Edge features should be numpy arrays"

    return GraphData(
        edge_index=graph_data.edge_index,
        node_features=graph_data.node_features,
        edge_features=graph_data.edge_features
    )


class SMILESDataset(Dataset):
    def __init__(self, df: pd.DataFrame, use_edges=False, use_partial_charge=False, allowable_set=None, cache_dir: Optional[str] = None):
        self.df = df
        self.graphs: list[GraphData] = []
        self.targets: list[np.ndarray] = []
        self.featurizer = MolGraphConvFeaturizer(use_edges=use_edges, use_partial_charge=use_partial_charge)
        self.allowable_set = allowable_set if allowable_set else ["C", "N", "O", "F", "P", "S", "Cl", "Br", "I", "other"]
        self.cache_dir = cache_dir
        self.is_for_dataloader = False
        super().__init__("custom_smiles_dataset")

    def download(self):
        pass

    def process(self):
        for idx, row in self.df.iterrows():
            smiles = row['smiles']
            target = row['target']

            cache_file = None
            if self.cache_dir:
                os.makedirs(self.cache_dir, exist_ok=True)
                cache_file = os.path.join(self.cache_dir, f"{smiles}.pkl")

            try:
                if cache_file and os.path.exists(cache_file):
                    graph_data = load_graph_data(cache_file)
                else:
                    graph_data = smiles_to_graph(smiles, self.featurizer, self.allowable_set)
                    if cache_file:
                        save_graph_data(graph_data, cache_file)

                self.graphs.append(graph_data)
                self.targets.append(np.array([target]))
            except Exception as e:
                print(f"Error processing SMILES '{smiles}': {e}")
                continue

    def __getitem__(self, idx: Union[int, np.integer, slice, mx.array, np.ndarray, Sequence]) -> Union["Dataset", GraphData]:
        indices = range(len(self))

        if isinstance(idx, (int, np.integer)) or (isinstance(idx, mx.array) and idx.ndim == 0):
            index = indices[idx]
            graph = self.graphs[index]

            if not isinstance(graph.node_features, np.ndarray):
                graph.node_features = np.array(graph.node_features)
            if graph.edge_features is not None and not isinstance(graph.edge_features, np.ndarray):
                graph.edge_features = np.array(graph.edge_features)

            if self.is_for_dataloader:
                return graph
            return graph, self.targets[index]

        if isinstance(idx, slice):
            indices = indices[idx]

        elif isinstance(idx, mx.array) and idx.dtype in [mx.int64, mx.int32, mx.int16]:
            return self[idx.flatten().tolist()]

        elif isinstance(idx, np.ndarray) and idx.dtype == np.int64:
            return self[idx.flatten().tolist()]

        elif isinstance(idx, Sequence) and not isinstance(idx, str):
            indices = [indices[i] for i in idx]

        else:
            raise IndexError(f"Dataset indexing failed. Accepted indices are: int, mx.array, list, tuple, np.ndarray.")

        dataset = copy.copy(self)
        graphs = [self.graphs[i] for i in indices]
        dataset.graphs = graphs
        dataset.targets = [self.targets[i] for i in indices]
        return dataset

    def set_for_dataloader(self, flag: bool):
        self.is_for_dataloader = flag


# Example usage:
if __name__ == "__main__":
    df = pd.DataFrame({
        'smiles': ['CCO', 'CCN', 'CCC', 'C1=CC=CC=C1'],
        'target': [0, 1, 0, 1]
    })

    dataset = SMILESDataset(df, use_edges=True, use_partial_charge=True, cache_dir="graph_cache")

    dataset.process()

    dataset.set_for_dataloader(True)

    train_loader = Dataloader(dataset, batch_size=2, shuffle=False)

    for batch in train_loader:
        print(batch.node_features)
        print(batch.edge_index)
        print(batch.edge_features)



```
**Describe alternatives you've considered**
Current code is not working this is just a proposal 


**Additional context**
Add any other context or screenshots about the feature request here.
